src/commonMain/kotlin/com/mattprecious/stacker/stack/tree.kt

```diff
@@ -0,0 +1,81 @@
+package com.mattprecious.stacker.stack
+
+fun <K : Any, T : Any> treeOf(
+	elements: Collection<T>,
+	keySelector: (T) -> K,
+	parentSelector: (T) -> K?,
+): TreeNode<T>? {
+	if (elements.isEmpty()) {
+		return null
+	}
+
+	val elementsByKey = elements.associateBy(keySelector)
+	val parents = elements.associateBy(keySelector) {
+		elements.indexOf(elementsByKey[parentSelector(it)])
+	}
+	val children = elements.groupBy(parentSelector) { elements.indexOf(it) }
+
+	require(parents.values.count { it == -1 } == 1) {
```

Comment:
I think this is `children[null]` on which you can then just do `.size` and also include the offending node list in the exception message.

Reply:
I had both of these checks in and removed the other one because it essentially asserting the same thing. Including it in the message is a good call though.